tools/cli/src/mentat_cli/command_parser.rs

```diff
@@ -90,34 +93,33 @@ pub fn command(s: &str) -> Result<Command, cli::Error> {
 
     let close_parser = string(CLOSE_COMMAND)
                     .with(arguments())
+                    .skip(spaces())
+                    .skip(eof())
                     .map(|args| {
                         if args.len() > 0 {
                             return Err(cli::ErrorKind::CommandParse(format!("Unrecognized argument {:?}", args[0])).into());
                         }
                         Ok(Command::Close)
                     });
     
-    let opening_brace_parser = try(string("["))
-                    .or(try(string("{")));
-    
-    let edn_arg_parser = spaces()
-                    .and(opening_brace_parser.clone()
-                    .and(many::<Vec<_>, _>(try(any()))));
-
-    let query_parser = string(QUERY_COMMAND)
-                    .and(edn_arg_parser.clone())
-                    .map( |x| {
-                        let args = (x.1).1;
-                        let content: String = args.1.iter().collect();
-                        Ok(Command::Query(format!("{}{}", args.0, content)))
-                    });
-
-    let transact_parser = string(TRANSACT_COMMAND)
-                    .and(edn_arg_parser.clone())
+    let edn_arg_parser = || spaces()
+                    .with(try(string("["))
+                        .or(try(string("{")))
+                        .then(|d| parser(move |input| {
```

Comment:
If I understand correctly, you're trying to parse any line prefixed by optional space and either `[` or `{`.  I think you can do that with:
```rust
spaces()
    .with(look_ahead(string("[").or(string("{")))
        .with(many(any()))
```
Using `then` and a special parser like you do is ingenious but definitely not a natural expression using `combine` :)  You might also do this as
```rust
spaces().then(many(any())).and_then(|line| {
    if line.starts_with("[") || line.starts_with("{") {
        Ok(...)
    } else {
        Err(...)  
    }
})
```
but then you need to know a lot more about error handling in `combine`.  Stick to the defined combinators for easiest use.